### PR TITLE
feat(test): add cpu time checks, clearer commands, nicer log

### PR DESCRIPTION
**Pull request by antoinemadec**
* commands.py now accepts for loops, if etc
* use raw strings in order to avoid double escaping

Both nvim and vim are passing.
I was however not able to run it in travis-ci, so you might wanna check this if your using travis-ci

### CODE REVIEW

**Comment by mg979** [comment]
I don't think travis CI is working (I'm not using it as you can see). I'll check the PR this evening, thanks!

**Comment by mg979** [comment]
I had a quick look, but just quick, because there's far too much stuff, both for a single commit and foe what I'm interested in. In short:

- changes to tests: no, totally unnecessary
- colors: it's nice but useless, I prefer to keep the test script slim because who's maintaining it then
- cpu time chek: I'm interested but if it doesn't break anything
- config.json: should be ok

I noticed that you removed the `for loops` that had a `sleep` inside them. They were there because they make the tests more reliable in some machines. I had to look well at the code to notice it, and such a big commit didn't make it easier.

Bottom line, I'm only interested in timers and the optional config.json, and only if the changes are kept minimal. I don't care about refactorings or prettifying anything. If these changes require a total rewrite of the script (like you almost did) I'm not interested. I don't mean to be blunt but time is short and that's too much stuff to handle, and even the potentially useful things are currently unnecessary.

**Comment by antoinemadec** [comment]
@mg979 , the changes to test make it clearer, I think using a single escaping is better than 2.
Do you prefer
```python
keys('\\\\\\\\a')
````
instead of 
```python
keys(r'\\\\a')
```
?

Colors make it nicer, if you think it makes the script too complex, feel free to remove them.
When running the test suite, color helped me identifying which test failed faster.

Cpu time does not break anything on my machine. The time built-in module functions I used are supposed to be cross plateform.
Do you have any issue with cpu time on your machine?

And finally, I did not remove the timings, I just reduced them (look for KEY_PRESS_INTERVAL, line 20).
Feel free to put it back to 0.3 sec instead of 0.1. It works fine on my 5 years old machine with both Vim and Neovim.  

I don't understand you remark about time being short.
The script is IMO clearer than before, and so are the commands.py (no more double escaping).
It looks better and is more future proof.

I already spent time to make the changes, and I had fun with it.
So, I don't really understand, but it's your repo so do what works for you.

Thanks.
Antoine

**Comment by antoinemadec** [comment]
Here is a clearer explanation about the need for raw strings.

1. perf tests needs **for loops** in **commands.py**
2. this implies sourcing **commands.py** directly instead of looping on each line
3. that implies keys=keys_vim where keys_vim implements both the feedkeys() and the sleep()
4. that means **2 indirections** (keys_vim() then feedkeys())  level instead of 1 previously (feedkeys() only)
5. those 2 indirections would have forced the keys strings to be **triple escaped** instead of **double escaped**

So in short, **for loops** implied changing the **commands.py**. No matter what.
It was either **triple escaped**:
```python
keys('\\\\\\\\\\\\\\\\a')
```
or raw strings, which does not remove escaping for each indirections:
```python
keys(r'\\\\a')
```

The structure of **test.py** has not changed besides this slight change, it should be readable for someone who understood the previous code.

The raw strings were not "totally unnecessary", there are needed to support for loops.
This was not a weird aesthetic flex of mine :smile: 

The other option is not to instrument the **for loops** and to copy paste blocks of code in commands.py.
This would have resulted in 50*4 = 200 lines of commands in my new testcase.

If this new code don't work on your machine, I understand that you don't want to implement it.
Otherwise, those improvements are... improvements and increase the test coverage.

I hope this makes things clearer!

**Comment by mg979** [comment]
> And finally, I did not remove the timings, I just reduced them

Ok I see now, sorry, I just looked at the commit and not at the whole file.

Overall I like it a lot now, but about raw strings, I still think they're not needed.

I tried the tests in their original form (both with vim and nvim), with the new `test.py`, and they run just as well. You say there is one more level of indirection but I don't think it changes anything. The string is passed as it is, whether it's raw or it's escaped inside of it, it doesn't matter, else tests would fail.

Basically only the VM leader (`\\`) would be shortened by using raw strings, and in most test scripts it's used a variable. Most `keys()` commands don't need to be escaped as raw strings, so that `r` is useless.

**Comment by antoinemadec** [comment]
> Overall I like it a lot now, but about raw strings, I still think they're not needed.

You are absolutely right.
I came to a point when I was developing where triple escaped was required, I assumed it was because of the 2 functions calls instead of one, I was wrong. Maybe it came from a wronge pattern in the replace() calls in keys_nvim?

So you were right, raw strings are not necessary.
However, now that it's done, considering that it looks better and is easier to use to create new tests, I think we should keep it.
If it's an ownership issue, don't hesitate to copy this commit and put you name on it. It is fine with me.

Sorry about the confusion and the BS about the function calls (I edited my previous answer).

Thanks,
Antoine

**Comment by mg979** [comment]
I don't want to amend the commit myself or you lose the contribution. I'd prefer that you amend it and just scrap the changes to tests and leave the changes to `test.py` alone. I can put that `r` in tests where it makes sense (and use the `L` variable where I'm not using it). I don't want to put the `r` everywhere because it's not needed.

**Comment by antoinemadec** [comment]
> I don't want to amend the commit myself or you lose the contribution. I'd prefer that you amend it and just scrap the changes to tests and leave the changes to `test.py` alone. I can then put that `r` in tests where it makes sense (and use the `L` variable where I'm not using it). I don't want to put the `r` everywhere because it's not needed.

I is not _needed_ anywhere, it is just more readable.
You could double the backslash everywhere.

How do you define a test where `r` makes sense?
Just so you know I won't spend effort on re-introducing double backslashes. This does not make sense to me.

**Comment by antoinemadec** [comment]
https://www.journaldev.com/23598/python-raw-string#:~:text=Python%20raw%20string%20is%20created,treated%20as%20an%20escape%20character.

**Comment by mg979** [comment]
An `r` makes only sense where there are double escapes (there aren't many because most of the times I use the `L` variable where there is the leader).

**Comment by antoinemadec** [comment]
Sorry, I won't spend time/effort to make the commands.py less consistent and readable.

The fact that you didn't want to change it in the first place, I understand.
The fact that once it is done you want to revert don't make sense to me.

As I was telling you before, I don't care about ownership, feel free to reuse this PR or not, copy/paste the code, revert what you want.
I did not modified the commands.py to have my name on it or to show off.
I thought it was needed, but wasn't. Now that the effort/work has been done, it does look better, is easier to mimic for a new test and is less confusing.

Spending time to revert it don't make sense, I do not understand your logic behind it.
I don't see any benefit of doing that.

Thanks for the great plugin anyway 👍

**Comment by mg979** [comment]
For me it's a 30 seconds thing.

- open test.py
- `:%y+`
- :!git reset --hard master^`
- `ggVG"+p`
- save, commit again and force push

And you keep the contribution.

**Comment by mg979** [comment]
And sorry for being stubborn, I just don't want unncessary stuff, it's a question of principle, if you like.

**Comment by antoinemadec** [comment]
> And sorry for being stubborn, I just don't want unnecessary stuff, it's a question of principle, if you like.

I respect your opinion but I don't share it 😃 
Functions are not necessary either, for loops are not necessary either. Factorizing KEY_PRESS_INTERVAL was not necessary either. Colors are not necessary either.
However, those are better practices which makes your life easier. It makes the code more readable and maintainable.

This is obviously not a time issue, as debating those changes is taking way more time than actually doing the changes 😄 
I guess you could call it a principle on my end too...
Now that the raw strings are hear, and that it is best practice, I don't want to create a commit to make the code _less good_.

Once again, your repo, your decision.
Cheers!

**Comment by mg979** [comment]
Ok no problem, I could amend the commit (excluding the changes to tests) and keep your authorship, I'll merge it.

**Comment by mg979** [comment]
Merged thanks, and sorry for the stubborness.

**Comment by balta2ar** [comment]
That was a very polite and respectful argument. You guys are cool.

**Comment by mg979** [comment]
If I knew before how to amend the commit while keeping the authorship I would have spared a part of the discussion, my bad. I didn't think it was possible but it's actually easy.

**Comment by antoinemadec** [comment]
> That was a very polite and respectful argument. You guys are cool.

Haha, agreed, we managed to disagree and to stay very cordial 😄 

I will try to reach out to the author of vim-multiple-cursor to ask him permission to modify his README.
I'd like to mention that it is currently unmaintained and add a link to vim-visual-multi.
Hopefully this will increase this plugin's following, which it clearly deserves.

Cheers!

**Comment by mg979** [comment]
@antoinemadec thanks but above all thanks for the work you've done, I hope you don't think I didn't appreciated it, and I must repeat that this plugin wouldn't have gone anywhere without the testing part you wrote (I would have totally burnt out for sure...).

**Comment by antoinemadec** [comment]
@mg979 , thanks a lot and no worries, it's actually great to be able to disagree calmly 😃
